snapcraft/plugins/jhbuild.py

```diff
@@ -0,0 +1,383 @@
+                     cwd=self.jhbuild_src)
+
+            self.run(['make', '-j%d' % self.parallel_build_count],
+                     cwd=self.jhbuild_src)
+
+            self.run(['make', '-j%d' % self.parallel_build_count,
+                      'install'], cwd=self.jhbuild_src)
+
+        archive_path = os.path.join(self.partdir, 'jhbuild', 'packages')
+        mirror_path = os.path.join(self.partdir, 'jhbuild', 'mirror')
+        unpacked_path = os.path.join(self.partdir, 'jhbuild', 'unpacked')
+
+        make_paths = [self.builddir, self.installdir,
+                      archive_path, mirror_path, unpacked_path]
+
+        for folder in make_paths:
+            os.makedirs(folder, exist_ok=True)
+
+        self._set_jhbuild_ownership(make_paths)
+
+    def _write_jhbuildrc(self):
+        """
+        _write_jhbuildrc() - write out the jhbuildrc file for build dependency
+        specification
+        """
+        with open(self.jhbuildrc_path, 'w') as jhbuildrc_file:
+            config = '''
+moduleset = {module_set!r}
+modulesets_dir = {modulesets_dir!r}
+tarballdir = {tarballdir!r}
+dvcs_mirror_dir = {dvcs_mirror_dir!r}
+checkoutroot = {checkoutroot!r}
+prefix = {prefix!r}
+buildroot = {buildroot!r}
+use_local_modulesets = True
+skip = [{skip!s}]
+module_autogenargs['gdk-pixbuf'] = '--disable-gio-sniffing'
+extra_prefixes = [{extra_prefixes!s}]
+cflags = {cflags!r}
+'''
+
+            extra_prefixes = [os.path.join(self.project.stage_dir, 'usr'),
+                              os.path.join(self.project.stage_dir, 'usr',
+                                           'local')]
+
+            jhbuildrc_file.write(
+                config.format(module_set=self.options.module_set,
+                              modulesets_dir=self.options.module_set_dir,
+                              tarballdir=self.options.jhbuild_archive or
+                              os.path.join(self.partdir, 'jhbuild',
+                                           'packages'),
+                              dvcs_mirror_dir=os.path.join(
+                                  self.partdir, 'jhbuild', 'mirror'),
+                              checkoutroot=os.path.join(
+                                  self.partdir, 'jhbuild', 'unpacked'),
+                              buildroot=os.path.join(self.builddir, 'jhbuild'),
+                              prefix=os.path.join(self.installdir, 'usr'),
+                              skip=', '.join(
+                                  ['\'%s\'' % module for module in
+                                   self.skip_modules]),
+                              extra_prefixes=', '.join(
+                                  ['\'%s\'' % prefix for prefix in
+                                   extra_prefixes]),
+                              cflags=self.options.cflags,
+                              ))
+
+    def pull(self):
+        self._pull_jhbuild()
+        self._setup_jhbuild()
```

Comment:
It's nicer when you are able to read the source code from top to bottom. For that, please move the definition of _pull_jhbuild, _setup_jhbuild and jhbuild after this pull method.
Then somebody new to the code will find first the pull method, and if he wants to dig further, he will also read the definitions of the other 3 methods. If he only wants a high level overview, he will just read pull. With the current implementation he will find first the definition of setup_jhbuild with no context to know if it's an implementation detail, or something critical to understand what the plugin does.